pkg/applications/providers/source/git.go

```diff
@@ -0,0 +1,217 @@
+
+		case appskubermaticv1.GitAuthMethodToken:
+			token, err := g.getCredentialFromSecret(credentials.Token.Name, credentials.Token.Key)
+			if err != nil {
+				return nil, err
+			}
+
+			auth = &http.TokenAuth{Token: token}
+		case appskubermaticv1.GitAuthMethodSSHKey:
+			privateKey, err := g.getCredentialFromSecret(credentials.SSHKey.Name, credentials.SSHKey.Key)
+			if err != nil {
+				return nil, err
+			}
+
+			authssh, err := ssh.NewPublicKeys("git", []byte(privateKey), "")
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse private ssh key: %w", err)
+			}
+			authssh.HostKeyCallback = gossh.InsecureIgnoreHostKey()
+
+			auth = authssh
+		default: // this should not happen.
+			return nil, fmt.Errorf("unknown Git authentication method '%s'", credentials.Method)
+		}
+	}
+	return auth, nil
+}
+
+func (g GitSource) getCredentialFromSecret(name string, key string) (string, error) {
+	secret := &corev1.Secret{}
+	if err := g.Client.Get(g.Ctx, types.NamespacedName{Namespace: g.SecretNamespace, Name: name}, secret); err != nil {
+		return "", fmt.Errorf("failed to get git credential: %w", err)
+	}
+
+	cred, found := secret.Data[key]
+	if !found {
+		return "", fmt.Errorf("key '%s' does not exist in secret '%s'", key, fmt.Sprintf("%s/%s", secret.GetNamespace(), secret.GetName()))
+	}
+	return string(cred), nil
+}
+
+// getCheckoutStrategy returns the checkoutFunc according to the GitSource.
+func (g GitSource) getCheckoutStrategy() checkoutFunc {
+	switch {
+	case len(g.Source.Ref.Commit) > 0:
+		return checkoutFromCommit
+
+	case len(g.Source.Ref.Branch) > 0:
+		return checkoutFromBranch
+
+	case len(g.Source.Ref.Tag) > 0:
+		return checkoutFromTag
+
+	default: // this should not happen.
+		return func(ctx context.Context, destination string, gitSource *appskubermaticv1.GitSource, auth transport.AuthMethod) error {
+			return fmt.Errorf("could not determine which reference to checkout")
+		}
+	}
+}
+
+// checkoutFunc define a function to clone and checkout code from repository defined in gitSource into destination using auth as credentials.
+type checkoutFunc func(ctx context.Context, destination string, gitSource *appskubermaticv1.GitSource, auth transport.AuthMethod) error
+
+// checkoutFromCommit clone the repository and checkout the desired commit.
+// If a branch is defined then a shallow clone of the branch is performed. The commit must belongs to this branch.
+func checkoutFromCommit(ctx context.Context, destination string, gitSource *appskubermaticv1.GitSource, auth transport.AuthMethod) error {
+	var repo *git.Repository
+	var err error
+	if len(gitSource.Ref.Branch) > 0 {
+		repo, err = git.PlainCloneContext(ctx, destination, false, &git.CloneOptions{
+			URL:           gitSource.Remote,
+			Auth:          auth,
+			Progress:      nil,
+			SingleBranch:  true,
+			ReferenceName: plumbing.NewBranchReferenceName(gitSource.Ref.Branch),
+			Tags:          git.NoTags,
+			NoCheckout:    true,
```

Comment:
Now that I see this again, I am wondering if this is not too complicated for the user. Essentially this creates a lot of possible states:
* branch only → super fast, depth 1 shallow clone
* branch+commit → medium fast, depth max, but only for one branch shallow clone
* commit only → full clone, either medium fast (if not too many other branches) or slow
* tag only → super fast, depth 1 shallow clone

In the end, this is due to git's design, which does not allow you to shallow clone a commit. And we know this, but the end-user doesn't. Therefore I think, we should at least re-word the user facing section here https://github.com/vgramer/kubermatic/blob/5c835b66f7315b1cbbcd6257a5e32f91a15a5c96/pkg/apis/apps.kubermatic/v1/application_definition.go#L84
to include the information.

Maybe we could do something like:
"For large repositories we recommend to either use Tag, Branch+Commit, or Branch. This allows a shallow clone, which dramatically speeds up performance"